Revamp troubleshooting docs

GiudGiud opened the pull request:
We always end up writing the same things in Discussions. It d be good to just link to the right section on this page for salvaging numerical solves

GiudGiud committed:
Add debug docs

GiudGiud committed:
Add mesh troubleshooting docs

GiudGiud committed:
Point to troubleshooting from debug
And from FAQ

GiudGiud committed:
Re-organize and expand troubleshooting documentation for each type of solve
refs #12466

GiudGiud committed:
Add controlling kernels option for initalization
Add how to diagnose linear vs nonlinear
Tell users to make a copy
Add NaN section

GiudGiud committed:
Add FAQ for profiling
Point to perf graph from profiling page for easy profiling

GiudGiud committed:
Add more initialization methods
Add jacobian troubleshooting
Re-organize sections

GiudGiud commented:
@fdkong I'd be interested in your opinions on this numerical troubleshooting page
Especially if I got something wrong!

Short link for your convenience
https://mooseframework.inl.gov/docs/PRs/24293/site/application_usage/failed_solves.html

GiudGiud commented:
@oanaoana as well

GiudGiud commented:
Will take a minute for the website to update with last commit

GiudGiud committed:
No need to flood output if empty or very common filename is used

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1521544/) on e830466 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>819845</th>
      <th colspan=4>#24293 e83046</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.20%</td>
      <td>85.20%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87717</td>
      <td>87714</td>
      <td>-3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15238</td>
      <td>15241</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24293/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24293/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24293/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Doco recipe fixes

fdkong commented:
> @fdkong I'd be interested in your opinions on this numerical troubleshooting page
Especially if I got something wrong!

> Short link for your convenience
https://mooseframework.inl.gov/docs/PRs/24293/site/application_usage/failed_solves.html

Will take a look into that this weekend. 

BTW why the tests failing?

GiudGiud commented:
Tests probably failing because either next is broken or I got a link wrong!

GiudGiud commented:
Patches to apps are made, not merged yet. 

I ll mark this as ready for review. Might be worth getting it in before the training?

GiudGiud commented on framework/doc/content/syntax/Mesh/index.md:
```diff
@@ -245,3 +245,36 @@ Mesh meta data can only be declared in the constructors of mesh generators so th
+It happens regularly that results are excellent and meeting all predictions using a regular Cartesian grid mesh,
+but significantly deteriorate or do not converge on the real system mesh, often created outside MOOSE.
+
+We point out in this section a few  to look for.
```
```suggestion
We point out in this section a few things to look for.
```

cticenhour commented on framework/doc/content/syntax/Debug/index.md:
```diff
@@ -71,6 +71,9 @@ these other helpful resources below:
   the [PETSc manual](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/).
 
+- to debug [Mesh](syntax/Mesh/index.md) related issue, please see this [page](syntax/Mesh/index.md#troubleshooting)
+  for in-MOOSE generation issues and this [page](syntax/Mesh/index.md#issues) if you suspect issues from an external mesh.
```
```suggestion
  for built-in MOOSE mesh generation issues and [this page](syntax/Mesh/index.md#issues) for possible issues from an external mesh.
```

cticenhour commented on framework/doc/content/syntax/Mesh/index.md:
```diff
@@ -74,7 +74,7 @@ mesh file with `_in.e` (the opposite of the `_out.e` that is appended from the o
 write out using `--mesh-only output_file.e`. When using the `--mesh-only` option, by default any extra element integers
 defined on the mesh will also be outputted to the output Exodus file. To prevent extra element ids from being
-outputted, the parameter `output_extra_element_ids` should be set to `false` in the `[Ouputs]` block of the
+outputted, the parameter `output_extra_element_ids` should be set to `false` in the `[Outputs]` block of the
```
```suggestion
output, the parameter `output_extra_element_ids` should be set to `false` in the `[Outputs]` block of the
```

cticenhour commented on framework/doc/content/syntax/Mesh/index.md:
```diff
@@ -245,3 +245,36 @@ Mesh meta data can only be declared in the constructors of mesh generators so th
+
+Mesh generation in MOOSE is a sequential tree-based process. Mesh generators are executed sorted by dependencies,
+and the output of each generator may be fed to multiple other generators. To succeed in this process, you must decompose
+the creation of the mesh in many individual steps. The two main tools to debug this process are:
```
```suggestion
the creation of the mesh into many individual steps. To debug this process, one can:
```

cticenhour commented on framework/doc/content/syntax/Mesh/index.md:
```diff
@@ -245,3 +245,36 @@ Mesh meta data can only be declared in the constructors of mesh generators so th
+and the output of each generator may be fed to multiple other generators. To succeed in this process, you must decompose
+the creation of the mesh in many individual steps. The two main tools to debug this process are:
+
+- use the `show_info(=true)` input parameter on each mesh generator. This will output numerous metadata about the mesh
```
```suggestion
- use the `show_info(=true)` input parameter on each mesh generator. This will output numerous pieces of metadata about the mesh
```

cticenhour commented on framework/doc/content/syntax/Mesh/index.md:
```diff
@@ -245,3 +245,36 @@ Mesh meta data can only be declared in the constructors of mesh generators so th
+but significantly deteriorate or do not converge on the real system mesh, often created outside MOOSE.
+
+We point out in this section a few  to look for.
+- sidesets in MOOSE are oriented. If you place a Neumann/flux boundary condition on a sideset, the direction of
```
```suggestion
- Sidesets in MOOSE are oriented. If you place a Neumann/flux boundary condition on a sideset, the direction of
```

GiudGiud committed:
Better words and syntax from Casey

Co-authored-by: Casey Icenhour <cticenhour@gmail.com>

moosebuild commented:
Job [Test heavy](https://civet.inl.gov/job/1557436/) on 5425d0b : invalidated by @GiudGiud

unrelated timeout

cticenhour commented:
@GiudGiud Blue-crab patch ready to go after this? Otherwise ready to merge.

GiudGiud commented:
I should update submodules indeed. I ll get to it.

Thanks for the review!
@fdkong @oanaoana I would still be happy to get your opinions on the troubleshooting failed solves page, but it can come as a patch to this as well.

lindsayad commented on modules/doc/content/application_usage/failed_solves.md:
```diff
@@ -67,44 +165,196 @@ Initial residual before setting preset BCs: 65444.1
+
+A more advanced check, and part of the process to fix the Jacobian, is to use the
+[Jacobian analysis utilities](development/analyze_jacobian.md). These will point out inaccurate or missing terms
+in the Jacobian. They can diagnose missing variable coupling terms for example. Once the errors have been identified,
```
```suggestion
in the Jacobian. They can diagnose missing variable coupling terms for example. Jacobian deficiencies for small problems may also be investigated from the command line using the PETSc options `-snes_test_jacobian -snes_test_jacobian_view`. Combined with [DOFMap.md] output, this can be similarly used to diagnose problematic areas of the Jacobian. Once the errors have been identified,
```

lindsayad commented on modules/doc/content/application_usage/failed_solves.md:
```diff
@@ -67,44 +165,196 @@ Initial residual before setting preset BCs: 65444.1
+But the memory cost will increase linearly with that number.
+Any zero eigenvalue will pose the same problem as a zero singular value.
+The size and values of eigenvalues will help you understand which preconditioner are likely to work, based on whether
+the problem is elliptic, parabolic or hyperbolic.
+
```
```suggestion
the problem is elliptic, parabolic or hyperbolic.

### Inaccurate Jacobian action

When using PJFNK with a noisy residual function, the Jacobian action computed via finite differencing of the residuals may be inaccurate. One may use the PETSc command line option `-ksp_monitor_true_residual` to see whether the true residual measured via Ax - b is decreasing (as opposed to the residual generated during GMRES iterations). If the GMRES residual is dropping but the true residual is not, then variables may need to be better scaled. Additionally the user may experiment with the `-mat_mffd_err` parameter. The default value is around 1e-8; however, values nearer 1e-6 or 1e-5 may be better for noisy functions in which a larger differencing may be necessary to cut through the noise.
```

GiudGiud commented:
Patch is ready.
I ll catch your comments in a follow up

GiudGiud committed:
Apply suggestions from code review
- add type for PJFNK with noisy residual
- add more tips on debugging Jacobian

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
might as well wait since next is broken

cticenhour commented:
You just hit MooseDocs whack-a-mole.....also Examples and tutorials failure is unrelated. I'm tweaking that recipe and goofed on a script.

GiudGiud commented:
Fixed

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1564238/) on f7d844d : invalidated by @lindsayad
